daemon/fqdn.go

```diff
@@ -73,11 +76,88 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
+				}
+			}
+
+			newPrefixLengths, err := d.prefixLengths.Add(ipNets)
+			if err != nil {
+				metrics.PolicyImportErrors.Inc()
+				log.WithError(err).WithField("prefixes", ipNets).Warn(
+					"Failed to reference-count prefix lengths in CIDR policy")
+				return err
+			}
+
+			if newPrefixLengths && !bpfIPCache.BackedByLPM() {
+				// Only recompile if configuration has changed.
+				log.Debug("CIDR policy has changed; recompiling base programs")
+				if err := d.compileBase(); err != nil {
+					_ = d.prefixLengths.Delete(ipNets)
+					err2 := fmt.Errorf("Unable to recompile base programs: %s", err)
+					log.WithError(err2).WithField("prefixes", ipNets).Warn(
+						"Failed to recompile base programs due to prefix length count change")
+					return err
+				}
+			}
+
+			// Used to track identities which are allocated in calls to
+			// AllocateCIDRs. If we for some reason cannot allocate new CIDRs,
+			// we have to undo all of our changes and release the identities.
+			// This is best effort, as releasing can fail as well.
+			usedIdentities := make([]*identity.Identity, 0)
+			selectorIdentitySliceMapping := make(map[string][]*identity.Identity)
+
+			// Allocate identities for each IPNet and then map to selector
+			for selector, selectorIPNets := range selectorIPNetMapping {
+				var currentlyAllocatedIdentities []*identity.Identity
+				if currentlyAllocatedIdentities, err = ipcache.AllocateCIDRs(bpfIPCache.IPCache, selectorIPNets); err != nil {
+					_ = d.prefixLengths.Delete(ipNets)
+					secIDCache.ReleaseSlice(context.Background(), usedIdentities)
+					log.WithError(err).WithField("prefixes", ipNets).Warn(
+						"Failed to allocate identities for CIDRs during policy add")
+					return err
+				}
+				usedIdentities = append(usedIdentities, currentlyAllocatedIdentities...)
+				selectorIdentitySliceMapping[selector] = currentlyAllocatedIdentities
+			}
+
+			// Update mapping of selector to set of IPs in selector cache.
+			for selector, identitySlice := range selectorIdentitySliceMapping {
+				log.WithFields(logrus.Fields{
+					"fqdnSelectorString": selector,
+					"identitySlice":      identitySlice}).Debug("updating FQDN selector")
+				numIds := make([]identity.NumericIdentity, len(identitySlice))
+				for _, numId := range identitySlice {
+					// Nil check here? Hopefully not necessary...
+					numIds = append(numIds, numId.ID)
+				}
+				policy.UpdateFQDNSelector(selector, numIds)
+			}
+
+			// Selectors which no longer map to IPs (due to TTL expiry, cache being
+			// cleared forcibly via CLI, etc.) still exist in the selector cache
+			// since policy is imported which allows it, but the selector does
+			// not match anything anymore.
+			for k := range selectorsWithoutIPs {
+				log.WithFields(logrus.Fields{
+					"fqdnSelectorString": selectorsWithoutIPs[k],
+				}).Debug("removing all identities from FQDN selector")
+				policy.UpdateFQDNSelector(selectorsWithoutIPs[k], []identity.NumericIdentity{})
```

Comment:
With [2] done you can call `policy.UpdateFQDNSelector` directly for all `selectorsWithoutIPs` at the same time.